### PR TITLE
feat(ethers-contract): Make ethers-providers optional

### PR DESCRIPTION
**Pull request by oblique**
## Motivation

I'm working in a very restricted WASM environment (see #2532), which I can not use `ethers-providers` crate. However I would like to use `ethers_contract::EthCall` derive functions.

## Solution

With this PR, I introduced the `providers` flag in `ethers-contract`, `ethers-contract-abigen`, and `ethers-contract-derive`. Anything related to `ethers-providers` is now enabled with the flag.

`providers` flag was added in the default features of `ethers-contract`.

If a user adds `ethers` or `ethers-middleware` crates, then `providers` is always enabled, even if user disabled the default flags. This is because they have `ethers-providers` as a dependency and there is no reason to make the flag optional.

If a user adds `ethers-contract` and disables default features, then they need to add `providers` feature to get the same functionality as before. - Is this edge case a major breaking change?

If you have any other ideas how this can be better, let me know.

## PR Checklist

-   [ ] Added Tests
-   [ ] Added Documentation
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-contract/src/multicall/middleware.rs
Comment (oblique) [inline ethers-contract/src/multicall/middleware.rs:1]: Content of `ethers-contract/src/multicall/mod.rs` was moved here

File: ethers-contract/src/event_core.rs
Comment (oblique) [inline ethers-contract/src/event_core.rs:1]: Content from `ethers-contract/src/event.rs` was moved here. Nothing else was changed.

**Comment by oblique** [comment]
The tests that are failing are not related to my changes and they failing before the changes on my machine.

**Comment by oblique** [comment]
I found that foundry was responsible for the test failures. The tests are passing with version `nightly-e05b9c75b4501d5880764948b61db787f3dd7fe0` and they fail with version `nightly-2d87c0c2fcc47088feecc72721c46d8e07e3c220`

**Comment by oblique** [comment]
With the latest foundry nightly the tests passed, so this PR is ready. Any comments are welcome.

File: ethers-contract/src/lib.rs [hunk 1]
```diff
@@ -3,24 +3,25 @@
 #![warn(missing_docs)]
 #![cfg_attr(docsrs, feature(doc_cfg))]
 
-#[path = "contract.rs"]
-mod _contract;
-pub use _contract::{Contract, ContractInstance};
+macro_rules! if_providers {
```
Comment (zvolin) [inline ethers-contract/src/lib.rs:6]: I'd prefer having `pub(crate) use` for this macro and importing/calling by full path on call site

Comment (oblique) [reply ethers-contract/src/lib.rs:6]: I want to avoid importing of this specific macro

File: ethers-contract/Cargo.toml [hunk 2]
```diff
@@ -23,9 +23,8 @@ rustdoc-args = ["--cfg", "docsrs"]
 all-features = true
 
 [dependencies]
-ethers-providers.workspace = true
+ethers-providers = { workspace = true, optional = true }
 ethers-core.workspace = true
-ethers-signers.workspace = true
```
Comment (DaniPopes) [inline ethers-contract/Cargo.toml:28]: Is this not needed anymore?

File: ethers-contract/ethers-contract-abigen/Cargo.toml [hunk 3]
```diff
@@ -50,6 +50,7 @@ tokio = { workspace = true, optional = true }
 url = { workspace = true, optional = true }
 
 [features]
+providers = []
```
Comment (DaniPopes) [inline ethers-contract/ethers-contract-abigen/Cargo.toml:53]: This should be enabled by default, likewise for `ethers-contract-derive`

**Review by DaniPopes — CHANGES_REQUESTED** [review]
This is great! Just a few things

File: ethers-contract/Cargo.toml (hunk 2 above)
Comment (oblique) [reply ethers-contract/Cargo.toml:28]: No, it is only used in `tests/it/abigen.rs`

File: ethers-contract/ethers-contract-abigen/Cargo.toml (hunk 3 above)
Comment (oblique) [reply ethers-contract/ethers-contract-abigen/Cargo.toml:55]: Done

**Review by DaniPopes — APPROVED** [review]
